### PR TITLE
Make operators with optional Tensor? arguments c10-full

### PR DESCRIPTION
**Pull request by smessmer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42043 Remove argument unpacking from variable kernels
* #42025 IValue stores undefined tensor as None
* #41947 Change C++ frontend to take optional<Tensor> arguments
* **#41610 Make operators with optional Tensor? arguments c10-full**

Previously, operators that have a `Tensor?` (i.e. optional tensor) in their schema implemented it using `Tensor` in C++ and filled in an undefined tensor for the None case.
The c10 operator library, however, expects `Tensor?` to be represented as `optional<Tensor>`, so those operators couldn't be c10-full yet and still had to use codegenerated unboxing instead of templated unboxing.

This PR changes that. It extends the `hacky_wrapper_for_legacy_signatures` to not only take care of TensorOptions, but now also map between signatures taking `Tensor` and `optional<Tensor>`.
For this, it requires an additional template parameter, the expected signature, and it uses that to go argument-by-argument and unwrap any optionals it finds.

Differential Revision: [D22607879](https://our.internmc.facebook.com/intern/diff/D22607879/)

### CODE REVIEW

File: aten/src/ATen/autocast_mode.cpp [hunk 1]
```diff
@@ -360,6 +360,10 @@ Therefore, for the moment, this is all copy pasted in from VariableTypeEverythin
   m.impl(REGISTER_NAME, \
     &WrapFunction<CastPolicy::POLICY, SIGNATURE, SIGNATURE, &FUNC>::type::call);
 
+#define KERNEL_HACKY_WRAPPER_FOR_LEGACY_SIGNATURES(FUNC, REGISTER_NAME, SIGNATURE, LEGACY_SIGNATURE, POLICY) \
```
Comment (bhosmer) [inline aten/src/ATen/autocast_mode.cpp:363]: Can I pitch `KERNEL_LEGACY_SIGNATURE`? I think that conveys the appropriate level of contingency, without being quite so much of an eyesore (intended, I know! :). It'd be one thing if these were all go away in a week, but it's not quite that ephemeral.

File: aten/src/ATen/autocast_mode.cpp [hunk 2]
```diff
@@ -360,6 +360,10 @@ Therefore, for the moment, this is all copy pasted in from VariableTypeEverythin
   m.impl(REGISTER_NAME, \
     &WrapFunction<CastPolicy::POLICY, SIGNATURE, SIGNATURE, &FUNC>::type::call);
 
+#define KERNEL_LEGACY_SIGNATURE(FUNC, REGISTER_NAME, SIGNATURE, LEGACY_SIGNATURE, POLICY) \
+  m.impl(REGISTER_NAME, \
+    c10::impl::hacky_wrapper_for_legacy_signatures<SIGNATURE>(TORCH_FN((&WrapFunction<CastPolicy::POLICY, LEGACY_SIGNATURE, LEGACY_SIGNATURE, &FUNC>::type::call))));
```
Comment (ezyang) [inline aten/src/ATen/autocast_mode.cpp:365]: I think this can be directly be eliminated. There aren't N functions you have to port: all you have to do is modify `cached_cast` to handle `optional<Tensor>`. Actually, I'm not sure if AMP properly supported undefined tensors previously anyway (cc @mcarilli )

File: aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h [hunk 3]
```diff
@@ -230,6 +230,22 @@ using supported_primitive_arg_types = guts::typelist::typelist<
       return ivalue_to_arg<std::vector<T>, AllowDeprecatedTypes>::call(std::move(v));
     }
   };
+  template<bool AllowDeprecatedTypes>
+  struct ivalue_to_arg<optional<ArrayRef<int64_t>>, AllowDeprecatedTypes> final {
+    // If an argument is optional<ArrayRef<int64_t>>, convert the IValue to a optional<std::vector<int64_t>> and pass that
+    // to the operator.
```
Comment (ezyang) [inline aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h:236]: Uhh, so I guess you're just following the same pattern as the code above (which was already doing this), but why do we have to convert the ivalue into a `std::vector<int64_t>`. Isn't this some totally unnecessary allocation?

File: aten/src/ATen/core/op_registration/hacky_wrapper_for_legacy_signatures.h [hunk 4]
```diff
@@ -83,13 +84,13 @@ struct with_scattered_tensor_options<UnderlyingFuncPtr, std::enable_if_t<has_ten
     using parameters_after_tensoroptions =
         guts::typelist::drop_t<gathered_parameter_types, tensoroptions_arg_index + 1>;
 
-    using wrapper = with_scattered_tensor_options_<UnderlyingFuncPtr, parameters_before_tensoroptions, parameters_after_tensoroptions>;
+    using wrapper = with_scattered_tensor_options__<UnderlyingFuncPtr, parameters_before_tensoroptions, parameters_after_tensoroptions>;
```
Comment (ezyang) [inline aten/src/ATen/core/op_registration/hacky_wrapper_for_legacy_signatures.h:87]: double underscore?!

File: torch/csrc/autograd/VariableTypeManual.cpp [hunk 5]
```diff
@@ -80,10 +80,11 @@ namespace {
 
 void backward(
     const Tensor& self,
-    const Tensor& gradient,
+    c10::optional<Tensor> gradient,
     c10::optional<bool> keep_graph,
     bool create_graph) {
-  torch::autograd::backward({self}, {gradient}, std::move(keep_graph), create_graph);
+  Tensor _gradient = gradient.has_value() ? *gradient : Tensor();
+  torch::autograd::backward({self}, {_gradient}, std::move(keep_graph), create_graph);
```
Comment (ezyang) [inline torch/csrc/autograd/VariableTypeManual.cpp:87]: Add a TODO to make `torch::autograd::backward` directly take `c10::optional` here?

File: aten/src/ATen/core/op_registration/hacky_wrapper_for_legacy_signatures.h (hunk 4 above)
Comment (ezyang) [reply aten/src/ATen/core/op_registration/hacky_wrapper_for_legacy_signatures.h:87]: (At this point it's probably worth saying why there's two levels of indirection here.)

**Review by ezyang — APPROVED** [review]
Approving to unblock.

File: aten/src/ATen/core/op_registration/hacky_wrapper_for_legacy_signatures.h (hunk 4 above)
Comment (bhosmer) [reply aten/src/ATen/core/op_registration/hacky_wrapper_for_legacy_signatures.h:87]: +1

File: aten/src/ATen/core/op_registration/hacky_wrapper_for_legacy_signatures.h [hunk 6]
```diff
@@ -105,11 +106,91 @@ struct with_scattered_tensor_options_<FuncPtr, guts::typelist::typelist<Paramete
...
+
+/**
+ * Take a kernel function that has a number of `Tensor` arguments
+ * and take in a `TargetSignature` that must match, but is allowed
+ * to take `optional<Tensor>` in place of some or all of the `Tensor`
+ * arguments. Returns a new kernel function that has `optional<Tensor>`
+ * in those locations, unwraps them to `Tensor` (potentially undefined tensor)
+ * and calls the original kernel function.
+ */
+template<class TargetSignature, class KernelFunc, std::enable_if_t<guts::typelist::true_for_any_type<is_optional_tensor_arg, typename guts::infer_function_traits_t<TargetSignature>::parameter_types>::value, int> = 0>
```
Comment (bhosmer) [inline aten/src/ATen/core/op_registration/hacky_wrapper_for_legacy_signatures.h:172]: how do you feel about line wrapping above like 100 or 120?

File: aten/src/ATen/core/op_registration/hacky_wrapper_for_legacy_signatures.h [hunk 7]
```diff
@@ -105,11 +106,91 @@ struct with_scattered_tensor_options_<FuncPtr, guts::typelist::typelist<Paramete
...
+constexpr bool _is_optional_tensor_arg() {
+    return std::is_same<c10::optional<at::Tensor>, std::decay_t<T>>::value;
+}
+template<class T> using is_optional_tensor_arg = guts::bool_constant<_is_optional_tensor_arg<T>()>;
+
+/**
+ * Take a kernel function that has a number of `Tensor` arguments
+ * and take in a `TargetSignature` that must match, but is allowed
+ * to take `optional<Tensor>` in place of some or all of the `Tensor`
+ * arguments. Returns a new kernel function that has `optional<Tensor>`
```
Comment (bhosmer) [inline aten/src/ATen/core/op_registration/hacky_wrapper_for_legacy_signatures.h:168]: Personally I'd really benefit from comments on each template definition describing what it does, in addition to a summary comment like this on the public endpoint. E.g. here, everything that's being described (other than the optional Tensor detection) is happening somewhere else, in uncommented templates. Doesn't have to be in depth, just something to label the pieces basically.

**Review by bhosmer — APPROVED** [review]
I'm finding the metaprogramming kind of hard to follow... some comment begging inline :)

File: aten/src/ATen/autocast_mode.cpp (hunk 2 above)
Comment (smessmer) [reply aten/src/ATen/autocast_mode.cpp:365]: this is eliminated in the next PR stacked on top.

File: aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h (hunk 3 above)
Comment (smessmer) [reply aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h:236]: This unboxes an `IValue` (which is a list, so it holds a `std::vector<IValue>`) to an `ArrayRef<int64_t>`. The only way of doing that is to copy the full list. I think I mentioned this a few times when I argued for not allowing `ArrayRef` in kernel signatures, but the resolution was that we said perf doesn't matter here.

File: aten/src/ATen/core/op_registration/hacky_wrapper_for_legacy_signatures.h (hunk 4 above)
Comment (smessmer) [reply aten/src/ATen/core/op_registration/hacky_wrapper_for_legacy_signatures.h:87]: I renamed it to `impl_`

File: aten/src/ATen/core/op_registration/hacky_wrapper_for_legacy_signatures.h (hunk 6 above)
Comment (smessmer) [reply aten/src/ATen/core/op_registration/hacky_wrapper_for_legacy_signatures.h:172]: generally in favor of clang-formatting everything but we haven't set that up yet

File: aten/src/ATen/core/op_registration/hacky_wrapper_for_legacy_signatures.h (hunk 7 above)
Comment (smessmer) [reply aten/src/ATen/core/op_registration/hacky_wrapper_for_legacy_signatures.h:168]: I added some more comments
